### ISSUE TITLE
Misleading error message when incompatible return type and `iter`

### ISSUE THREAD

**Issue by Viicos** [python/mypy#15408]
**Bug Report**

I was adding type hints to an `__iter__` method which returns an `Iterator` from a `tuple` with different types. MRE is as follows:

```python
def func(cls: type, **options: Any) -> Iterator[Union[type, Dict[str, Any]]]:
    as_tuple = (cls, options)
    return iter(as_tuple)
```

As you can see `Iterator[Union[type, Dict[str, Any]]]` is wrong with `mypy` because of how it handles [unions vs joins](https://microsoft.github.io/pyright/#/mypy-comparison?id=unions-vs-joins) (`Iterator[object]` works).

But the problem I had is the error message returned by this snippet:

```sh
error: Argument 1 to "iter" has incompatible type "tuple[type, dict[str, Any]]"; expected "SupportsIter[Iterator[type | dict[str, Any]]]"  [arg-type]
```

While when using:

```python
def func(cls: type, **options: Any) -> Iterator[Union[type, Dict[str, Any]]]:
    as_tuple = (cls, options)
    rv = iter(as_tuple)
    return rv
```

We get a way better error message:

```sh
error: Incompatible return value type (got "Iterator[object]", expected "Iterator[type | dict[str, Any]]")  [return-value]
```
I don't know if this is specific to this `iter` use case, but never encountered it before.

- Mypy version used: master
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.11
